### PR TITLE
MGMT-1727: subsystem tests for auth/authz using WireMock

### PR DESCRIPTION
**Pull request by danielerez**
Added subsystem tests for authentication/authorization using WireMock:
* Added a deployment a service - for WireMock (using rodolpheche/wiremock image).
* Added wiremock_stubs.go - for creating stubs of the following OCM APIs:
  * AuthenticatePullSecret
  * AccessReview
  * CapabilityReview
* Added some negative tests to: host_test.go and authz_test.go

### CODE REVIEW

File: deploy/wiremock/wiremock-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      containers:
+        - name: wiremock
+          image: rodolpheche/wiremock
+          imagePullPolicy: "IfNotPresent"
+          ports:
+            - containerPort: 8080
+          volumeMounts:
+            - mountPath: /home/wiremock
+              name: wiremock
+      volumes:
```
Comment (avishayt) [inline deploy/wiremock/wiremock-deployment.yaml:25]: Why does wiremock need a persistent volume?  Isn't it all ephemeral?

**Review by avishayt — COMMENTED** [review]
Overall looks good.  One comment inline.  I think we should add tests for more APIs, but that can be later.  We also have the admin role that can view all clusters, right?

**Comment by danielerez** [comment]
> Overall looks good. One comment inline. I think we should add tests for more APIs, but that can be later. We also have the admin role that can view all clusters, right?

right, add an explicit test for this. could indeed test more APIs later on.

File: deploy/wiremock/wiremock-deployment.yaml (hunk 1 above)
Comment (danielerez) [reply deploy/wiremock/wiremock-deployment.yaml]: right, we actually don't use any static stubs, so can remove the volume.

**Comment by danielerez** [comment]
@avishayt addressed the comments, can you please take another look?

**Comment by danielerez** [comment]
@avishayt addressed the comments, can you please take another look?

File: Makefile [hunk 2]
```diff
@@ -173,11 +173,11 @@ deploy-postgres: deploy-namespace
 	python3 ./tools/deploy_postgres.py --namespace "$(NAMESPACE)" --profile "$(PROFILE)" --target "$(TARGET)"
 
 jenkins-deploy-for-subsystem: generate-keys build-dummy-ignition-image build-assisted-iso-generator-image
-	export TEST_FLAGS=--subsystem-test && export ENABLE_AUTH="True" && export DUMMY_IGNITION=${DUMMY_IGNITION} && $(MAKE) deploy-all
+	export TEST_FLAGS=--subsystem-test && export ENABLE_AUTH="True" && export DUMMY_IGNITION=${DUMMY_IGNITION} && $(MAKE) deploy-wiremock deploy-all
 
 deploy-test: generate-keys
 	export SERVICE=minikube-local-registry/assisted-service:minikube-test && export TEST_FLAGS=--subsystem-test && export ENABLE_AUTH="True" \
-	&& export DUMMY_IGNITION=${DUMMY_IGNITION} && ISO_CREATION=minikube-local-registry/assisted-iso-create:minikube-test $(MAKE) update-minikube deploy-all
+	&& export DUMMY_IGNITION=${DUMMY_IGNITION} && ISO_CREATION=minikube-local-registry/assisted-iso-create:minikube-test $(MAKE) update-minikube deploy-wiremock deploy-all
```
Comment (filanov) [inline Makefile:180]: please break long lines.

File: Makefile [hunk 3]
```diff
@@ -195,10 +195,16 @@ test:
 	INVENTORY=$(shell $(call get_service,assisted-service) | sed 's/http:\/\///g') \
 		DB_HOST=$(shell $(call get_service,postgres) | sed 's/http:\/\///g' | cut -d ":" -f 1) \
 		DB_PORT=$(shell $(call get_service,postgres) | sed 's/http:\/\///g' | cut -d ":" -f 2) \
+		OCM_HOST=$(shell $(call get_service,wiremock) | sed 's/http:\/\///g') \
```
Comment (filanov) [inline Makefile:198]: please make sure that subsystem can run onprem deployment as well - they are using the same subsystem.

File: Makefile [hunk 4]
```diff
@@ -195,10 +195,16 @@ test:
...
 		TEST_TOKEN="$(shell cat $(BUILD_FOLDER)/auth-tokenString)" \
+		TEST_TOKEN_ADMIN="$(shell cat $(BUILD_FOLDER)/auth-tokenAdminString)" \
+		TEST_TOKEN_UNALLOWED="$(shell cat $(BUILD_FOLDER)/auth-tokenUnallowedString)" \
 		ENABLE_AUTH="true" \
 		go test -v ./subsystem/... -count=1 -ginkgo.focus=${FOCUS} -ginkgo.v -timeout 30m
 
+deploy-wiremock: deploy-namespace
+	python3 ./tools/deploy_wiremock.py --target $(TARGET)
```
Comment (filanov) [inline Makefile:206]: why whiremock need a target parameter?

Comment (filanov) [inline Makefile:206]: namespace probably should be a parameter.

File: subsystem/authz_test.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/openshift/assisted-service/client"
+	"github.com/openshift/assisted-service/client/installer"
+	"github.com/openshift/assisted-service/models"
+)
+
+var _ = Describe("test authorization", func() {
```
Comment (filanov) [inline subsystem/authz_test.go:15]: Need to make sure that it will work with onprem

File: subsystem/subsystem_suite_test.go [hunk 6]
```diff
@@ -53,16 +56,35 @@ func init() {
...
 			Options.DBHost, Options.DBPort))
 	if err != nil {
 		logrus.Fatal("Fail to connect to DB, ", err)
 	}
+
+	var wiremock = &WireMock{
+		OCMHost:   Options.OCMHost,
+		TestToken: Options.TestToken,
+	}
+	err = wiremock.DeleteAllWiremockStubs()
```
Comment (ori-amizur) [inline subsystem/subsystem_suite_test.go:80]: Seems strange that the stubs are installed as part of the clientcfg function.  It should be in context of the tests you wish to perform.  Shouldn't it be part of BeforeEach or similar?

File: pkg/auth/authz_handler.go [hunk 7]
```diff
@@ -52,9 +50,6 @@ func (a *AuthzHandler) CreateAuthorizer() func(*http.Request) error {
 // Authorizer is used to authorize a request after the Auth function was called using the "Auth*" functions
 // and the principal was stored in the context in the "AuthKey" context value.
 func (a *AuthzHandler) Authorizer(request *http.Request) error {
-	if !a.EnableAuth {
```
Comment (nmagnezi) [inline pkg/auth/authz_handler.go:55]: why was this removed?

File: Makefile (hunk 4 above)
Comment (nmagnezi) [reply Makefile:206]: +1

File: tools/deploy_wiremock.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import utils
+import deployment_options
+import pvc_size_utils
+
+
+log = utils.get_logger('deploy_wiremock')
+
+
+def main():
```
Comment (nmagnezi) [inline tools/deploy_wiremock.py:9]: why leave an blank line between each line?

File: tools/deploy_wiremock.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    log.info('Starting wiremock deployment')
+
+    utils.set_profile(deploy_options.target, deploy_options.profile)
+
+    deploy_wiremock(deploy_options)
+
+    log.info('Completed to wiremock deployment')
+
+
+def deploy_wiremock(deploy_options):
```
Comment (nmagnezi) [inline tools/deploy_wiremock.py:21]: ditto (blank lines)

File: tools/deploy_wiremock.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    docs = utils.load_yaml_file_docs('deploy/wiremock/wiremock-deployment.yaml')
+
+    utils.set_namespace_in_yaml_docs(docs, deploy_options.namespace)
+
+    dst_file = utils.dump_yaml_file_docs('build/wiremock-deployment.yaml', docs)
```
Comment (nmagnezi) [inline tools/deploy_wiremock.py:26]: should this catch exceptions and fail in those cases? It does open files and more..

File: tools/auth_keys_generator.go [hunk 11]
```diff
@@ -71,7 +74,7 @@ func fileExists(filename string) bool {
 }
 
 func newFile(filename string, data []byte, perm os.FileMode) error {
-	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_EXCL, perm)
+	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, perm)
```
Comment (nmagnezi) [inline tools/auth_keys_generator.go:77]: please explain why.
Isn't `O_EXCL` what we want? 
```golang
O_EXCL   int = syscall.O_EXCL   // used with O_CREATE, file must not exist.
```
([source](https://golang.org/pkg/os/#pkg-constants))

File: pkg/auth/authz_handler.go (hunk 7 above)
Comment (danielerez) [reply pkg/auth/authz_handler.go:55]: this was just a workaround to make the subsystem work. now that we have wiremock we can actually test the flows :) so we simply create an empty authorizer in CreateAuthorizer in case EnableAuth is false.

File: tools/deploy_wiremock.py (hunk 8 above)
Comment (danielerez) [reply tools/deploy_wiremock.py:8]: just an 'inspiration' from similar deploy files ;-) can remove.

File: tools/deploy_wiremock.py (hunk 9 above)
Comment (danielerez) [reply tools/deploy_wiremock.py:15]: yeah, seems it's a convention in the deploy files to keep two blank lines between funcs.

File: tools/deploy_wiremock.py (hunk 10 above)
Comment (danielerez) [reply tools/deploy_wiremock.py:18]: hmm, haven't see any special handling in the other deploy scripts. but good point, maybe worth handling for all scripts.

File: tools/auth_keys_generator.go (hunk 11 above)
Comment (danielerez) [reply tools/auth_keys_generator.go:77]: Wanted to support updating the files as we added some new tokens (auth-tokenAdminString/auth-tokenUnallowedString). So if a user already generated keys, instead of failing it would update the tokens and create the new ones. Do you think it's better to fail and ask the user to run 'make clean' in such case?

Comment (nmagnezi) [reply tools/auth_keys_generator.go:77]: IMO, files override should be deliberate.

This currently prevents a case in which a tester loads a local test cert to assisted-service, but later on, overrides the token they need to use in order to query the API.

This is why I wrote it to fail like that.
Nevertheless, a `"--force"` flag can be added to this keys generator so we'll know the regeneration is done by purpose.

File: pkg/auth/authz_handler.go (hunk 7 above)
Comment (nmagnezi) [reply pkg/auth/authz_handler.go:55]: I see.
Just wondered because we do [refrain from full initialization](https://github.com/openshift/assisted-service/blob/6b39652ff43206e1e8957742ded7861367182f5b/pkg/auth/auth_handler.go#L49 
) of AuthHandler (aka empty AuthHandler) when `EnableAuth` is set to `False` (also [checked upon incoming requests](https://github.com/openshift/assisted-service/blob/6b39652ff43206e1e8957742ded7861367182f5b/pkg/auth/auth_handler.go#L222)).

Can / should we do the same here?

Comment (danielerez) [reply pkg/auth/authz_handler.go:55]: Yeah, we do something similar for authz here: https://github.com/openshift/assisted-service/blob/6b39652ff43206e1e8957742ded7861367182f5b/pkg/auth/authz_handler.go#L43

File: tools/auth_keys_generator.go [hunk 12]
```diff
@@ -84,3 +87,62 @@ func newFile(filename string, data []byte, perm os.FileMode) error {
...
+		"last_name":      "Doe",
+		"type":           "User",
+		"locale":         "en_US",
+		"first_name":     "John",
+		"email":          "unallowed@example.com",
+		"username":       "unallowed@example.com",
+		"is_org_admin":   false,
+		"clientId":       "1234",
+	})
+}
```
[the remark this answers is not in the document]
Comment (danielerez) [reply tools/auth_keys_generator.go]: sure, refactored a bit :)

File: tools/auth_keys_generator.go (hunk 11 above)
Comment (danielerez) [reply tools/auth_keys_generator.go]: ok, removed the override support. users should simply invoke clean manually to remove stale tokens.

**Comment by danielerez** [comment]
@filanov @ori-amizur @nmagnezi thanks for the review :) addressed comments, can you please take another look?

**Review by nmagnezi — CHANGES_REQUESTED** [review]
subsystem tests fail because it cannot find the token you generate

@danielerez 
```
[2020-09-07T08:38:34.749Z] + make subsystem-run
[2020-09-07T08:38:35.686Z] cat: /home/workspace/assisted-service_PR-203/build/assisted-installer/auth-tokenString: No such file or directory
[2020-09-07T08:38:35.687Z] cat: /home/workspace/assisted-service_PR-203/build/assisted-installer/auth-tokenAdminString: No such file or directory
[2020-09-07T08:38:35.687Z] cat: /home/workspace/assisted-service_PR-203/build/assisted-installer/auth-tokenUnallowedString: No such file or directory
```

**Comment by danielerez** [comment]
> subsystem tests fail because it cannot find the tokens you generate
> 
> @danielerez
> 
> ```
> [2020-09-07T08:38:34.749Z] + make subsystem-run
> [2020-09-07T08:38:35.686Z] cat: /home/workspace/assisted-service_PR-203/build/assisted-installer/auth-tokenString: No such file or directory
> [2020-09-07T08:38:35.687Z] cat: /home/workspace/assisted-service_PR-203/build/assisted-installer/auth-tokenAdminString: No such file or directory
> [2020-09-07T08:38:35.687Z] cat: /home/workspace/assisted-service_PR-203/build/assisted-installer/auth-tokenUnallowedString: No such file or directory
> ```

It was a temporary issue in subsystem cleanup, CI passes now :)

**Review by filanov — APPROVED** [review]

**Comment by danielerez** [comment]
> New changes are detected. LGTM label has been removed.

rebased

File: pkg/auth/authz_handler.go (hunk 7 above)
Comment (nmagnezi) [reply pkg/auth/authz_handler.go:55]: ack, looks good.

File: tools/deploy_wiremock.py (hunk 10 above)
Comment (nmagnezi) [reply tools/deploy_wiremock.py:18]: ack. this can happen in a different PR.

**Comment by nmagnezi** [comment]
This got merged by the time I finished 😄 , but in any case:
/lgtm
